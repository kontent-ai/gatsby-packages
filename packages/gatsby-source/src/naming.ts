import { PluginNamingConfiguration } from './types';
import { ItemsSchemaTemplate, TaxonomiesSchemaTemplate, TypesSchemaTemplate } from "./schemaTemplates";

const CONNECTOR = '_';
const SYSTEM_IDENTIFIER = 'system';
const ELEMENT_IDENTIFIER = 'element';
const VALUE_IDENTIFIER = 'value';
const TAXONOMY_TERM_IDENTIFIER = 'term';
const ELEMENT_OPTION_IDENTIFIER = 'option';
const MULTI_ELEMENT_IDENTIFIER = `${ELEMENT_IDENTIFIER}s`;
const LANGUAGE_LINK_EXTENSION_IDENTIFIER = 'language_link';
const ITEM_IDENTIFIER = 'item';
const TAXONOMY_IDENTIFIER = 'taxonomy';
const TYPE_IDENTIFIER = 'type';
const CACHE_IDENTIFIER = 'cache';

const RICH_TEXT_ELEMENT_TYPE_NAME = 'rich_text';

const PREFERRED_LANGUAGE_IDENTIFIER = 'preferred_language';

const defaultPluginNamingConfiguration: PluginNamingConfiguration = {
  prefix: `kontent`,
};

/**
 * Retrieve ID string for Gatsby's CreateNodeId method for specified item language variant.
 * @param id  Item `system.id` because codename is changeable.
 * @param preferredLanguage Preferred language of the language variant.
 * @param config Optional parameter with extra configuration.
 */
const getKontentItemNodeStringForId = (
  id: string,
  preferredLanguage: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${preferredLanguage}${CONNECTOR}${id}`;

const getKontentItemNodeTypeName = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}`;

const getKontentItemInterfaceName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}`;

const getKontentItemSystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentItemElementTypeNameByType = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${ELEMENT_IDENTIFIER}${CONNECTOR}${VALUE_IDENTIFIER}`;

const getKontentItemElementValueTypeNameByType = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${ELEMENT_IDENTIFIER}`;

const getKontentItemElementsSchemaTypeName = (
  type: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${type}${CONNECTOR}${MULTI_ELEMENT_IDENTIFIER}`;

const getKontentItemLanguageLinkExtensionName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${ITEM_IDENTIFIER}${CONNECTOR}${LANGUAGE_LINK_EXTENSION_IDENTIFIER}`;

const getKontentTaxonomyNodeStringForCodeName = (
  codename: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${codename}`;

const getKontentTaxonomyTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}`;

const getKontentTaxonomySystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentTaxonomyTermTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TAXONOMY_IDENTIFIER}${CONNECTOR}${TAXONOMY_TERM_IDENTIFIER}`;

const getKontentTypeNodeStringForCodeName = (
  codename: string,
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${codename}`;

const getKontentTypeTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}`;

const getKontentTypeSystemElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${SYSTEM_IDENTIFIER}`;

const getKontentTypeElementTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${ELEMENT_IDENTIFIER}`;

const getKontentTypeElementOptionTypeName = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${getKontentTypeElementTypeName(
    config,
  )}${CONNECTOR}${ELEMENT_OPTION_IDENTIFIER}`;

const getKontentTypesCacheKey = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string =>
  `${config.prefix}${CONNECTOR}${TYPE_IDENTIFIER}${CONNECTOR}${CACHE_IDENTIFIER}`;

const getKontentItemsSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return (
    ItemsSchemaTemplate
      .replace(
        /__KONTENT_ITEM_INTERFACE__/g,
        getKontentItemInterfaceName(config),
      )
      .replace(
        /__KONTENT_ITEM_SYSTEM_TYPE__/g,
        getKontentItemSystemElementTypeName(config),
      )
      // elements
      .replace(
        /__KONTENT_ITEM_TEXT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('text', config),
      )
      .replace(
        /__KONTENT_ITEM_RICH_TEXT_ELEMENT__/g,
        getKontentItemElementTypeNameByType(RICH_TEXT_ELEMENT_TYPE_NAME, config),
      )
      .replace(
        /__KONTENT_ITEM_NUMBER_ELEMENT__/g,
        getKontentItemElementTypeNameByType('number', config),
      )
      .replace(
        /__KONTENT_ITEM_MULTIPLE_CHOICE_ELEMENT__/g,
        getKontentItemElementTypeNameByType('multiple_choice', config),
      )
      .replace(
        /__KONTENT_ITEM_DATE_TIME_ELEMENT__/g,
        getKontentItemElementTypeNameByType('date_time', config),
      )
      .replace(
        /__KONTENT_ITEM_ASSET_ELEMENT__/g,
        getKontentItemElementTypeNameByType('asset', config),
      )
      .replace(
        /__KONTENT_ITEM_MODULAR_CONTENT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('modular_content', config),
      )
      .replace(
        /__KONTENT_ITEM_CUSTOM_ELEMENT_ELEMENT__/g,
        getKontentItemElementTypeNameByType('custom', config),
      )
      .replace(
        /__KONTENT_ITEM_TAXONOMY_ELEMENT__/g,
        getKontentItemElementTypeNameByType('taxonomy', config),
      )
      .replace(
        /__KONTENT_ITEM_URL_SLUG_ELEMENT__/g,
        getKontentItemElementTypeNameByType('url_slug', config),
      )
      // element values
      .replace(
        /__KONTENT_ELEMENT_MULTIPLE_CHOICE_VALUE__/g,
        getKontentItemElementValueTypeNameByType('multiple_choice', config),
      )
      .replace(
        /__KONTENT_ELEMENT_ASSET_VALUE__/g,
        getKontentItemElementValueTypeNameByType('asset', config),
      )
      .replace(
        /__KONTENT_ELEMENT_TAXONOMY_VALUE__/g,
        getKontentItemElementValueTypeNameByType('taxonomy', config),
      )
      .replace(
        /__KONTENT_ELEMENT_RICH_TEXT_IMAGE_VALUE__/g,
        `${getKontentItemElementValueTypeNameByType(
          RICH_TEXT_ELEMENT_TYPE_NAME,
          config,
        )}${CONNECTOR}image`,
      )
      .replace(
        /__KONTENT_ELEMENT_RICH_TEXT_LINK_VALUE__/g,
        `${getKontentItemElementValueTypeNameByType(
          RICH_TEXT_ELEMENT_TYPE_NAME,
          config,
        )}${CONNECTOR}link`,
      )
      // extensions
      .replace(
        /__KONTENT_ITEM_LANGUAGE_EXTENSION__/g,
        getKontentItemLanguageLinkExtensionName(config),
      )
      .replace(
        /__KONTENT_ITEM_PREFERRED_LANGUAGE_IDENTIFIER__/g,
        PREFERRED_LANGUAGE_IDENTIFIER,
      )
  );
};

const getKontentTaxonomiesSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return TaxonomiesSchemaTemplate
    .replace(/__KONTENT_TAXONOMY_NAME__/g, getKontentTaxonomyTypeName(config))
    .replace(
      /__KONTENT_TAXONOMY_SYSTEM_TYPE__/g,
      getKontentTaxonomySystemElementTypeName(config),
    )
    .replace(
      /__KONTENT_TAXONOMY_TERM_TYPE__/g,
      getKontentTaxonomyTermTypeName(config),
    );
};

const getKontentTypesSchemaNamingConfiguration = (
  config: PluginNamingConfiguration = defaultPluginNamingConfiguration,
): string => {
  return TypesSchemaTemplate
    .replace(/__KONTENT_TYPE_NAME__/g, getKontentTypeTypeName(config))
    .replace(
      /__KONTENT_TYPE_SYSTEM_TYPE__/g,
      getKontentTypeSystemElementTypeName(config),
    )
    .replace(
      /__KONTENT_TYPE_ELEMENT_TYPE__/g,
      getKontentTypeElementTypeName(config),
    )
    .replace(
      /__KONTENT_TYPE_ELEMENT_OPTIONS_TYPE__/g,
      getKontentTypeElementOptionTypeName(config),
    );
};

export {
  getKontentItemNodeStringForId,
  getKontentItemNodeTypeName,
  getKontentItemSystemElementTypeName,
  getKontentItemInterfaceName,
  getKontentItemElementTypeNameByType,
  getKontentItemElementsSchemaTypeName,
  getKontentItemLanguageLinkExtensionName,
  getKontentTaxonomyNodeStringForCodeName,
  getKontentTaxonomyTypeName,
  getKontentTypeNodeStringForCodeName,
  getKontentTypeTypeName,
  getKontentTypesCacheKey,
  getKontentItemsSchemaNamingConfiguration,
  getKontentTaxonomiesSchemaNamingConfiguration,
  getKontentTypesSchemaNamingConfiguration,
  PREFERRED_LANGUAGE_IDENTIFIER,
  RICH_TEXT_ELEMENT_TYPE_NAME
};
