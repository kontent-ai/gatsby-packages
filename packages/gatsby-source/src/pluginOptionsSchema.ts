
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const pluginOptionsSchema = ({ Joi }: { Joi: any }): any => {
  return Joi.object({
    projectId: Joi.string()
      .required()
      .length(36)
      .description(`Project ID from "Project settings" -> "API keys".`),
    languageCodenames: Joi.array().items(Joi.string())
      .required()
      .min(1)
      .description(`Array of language codenames from "Project settings" -> "Localization" - the first one is considered as the default one.`),
    includeTypes: Joi.boolean()
      .description(`Include content type nodes to GraphQL model.`)
      .default(false),
    includeTaxonomies: Joi.boolean()
      .description(`Include taxonomy node to GraphQL model.`)
      .default(false),
    authorizationKey: Joi.string()
      .description(`For preview/secured API key - depends on "usePreviewUrl" config.`),
    usePreviewUrl: Joi.boolean()
      .description(`When true, "preview-deliver.kontent.ai" used as primary domain for data source.`)
      .default(false),
    proxy: Joi.object()
      .keys({
        deliveryDomain: Joi.string()
          .description(`Base url used for all requests. Defaults to "deliver.kontent.ai".`),
        previewDeliveryDomain: Joi.string()
          .description(`Base url used for preview requests. Defaults to "preview-deliver.kontent.ai".`)
      })
      .description(`Proxy domain setting object`),
    includeRawContent: Joi.boolean()
      .description("Include `internal.content` property as a part fo the GraphQL model.")
      .default(false),
    experimental: Joi.object()
    .keys({
      managementApiTriggersUpdate: Joi.boolean()
      .description("Turn on experimental handling of management API triggers for update.")
      .default(false),
      additionalItemFilterParams : Joi.string()
      .description("Additional item filter parameters to reduce the content in GraphQL model.")
      .default("")
    })
  });
}

export {
  pluginOptionsSchema
}