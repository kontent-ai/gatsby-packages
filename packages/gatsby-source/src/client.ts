import * as rax from 'retry-axios';
import axios, { AxiosError } from 'axios';
import {
  KontentItemInput,
  KontentType,
  KontentTaxonomy,
  CustomPluginOptions,
} from './types';
import * as _ from 'lodash';
import {
  name as packageName,
  version as packageVersion,
} from '../package.json';
import { GatsbyCache } from 'gatsby';
import { getKontentTypesCacheKey } from './naming';

const DefaultKontentDeliveryProductionDomain = 'deliver.kontent.ai';
const DefaultKontentDeliveryPreviewDomain = 'preview-deliver.kontent.ai';
const continuationHeaderName = 'x-continuation';
const authorizationHeaderName = 'authorization';
const trackingHeaderName = 'x-kc-source';
const waitForLoadingNewContentHeaderName = 'x-kc-wait-for-loading-new-content';

rax.attach();

const getProtocolAndDomain = (options: CustomPluginOptions): string => {
  const domain = options.usePreviewUrl
    ? options?.proxy?.previewDeliveryDomain || DefaultKontentDeliveryPreviewDomain
    : options?.proxy?.deliveryDomain || DefaultKontentDeliveryProductionDomain;
  return `https://${domain}`;
};

const logRetryAttempt = (err: AxiosError): void => {
  const cfg = rax.getConfig(err);
  
  console.log(
    `Error axios request:(url: ${err.response?.config.url}) ${err.message}`,
  );
  console.log(`Retry attempt #${cfg?.currentRetryAttempt}`);
};

interface KontentHttpHeaders {
  [continuationHeaderName]?: string;
  [authorizationHeaderName]?: string;
  [trackingHeaderName]?: string;
  [waitForLoadingNewContentHeaderName]?: string;
}

const ensureAuthorizationHeader = (
  config: CustomPluginOptions,
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  if (!config.authorizationKey) {
    return headers || {};
  }

  if (headers) {
    headers[authorizationHeaderName] = `Bearer ${config.authorizationKey}`;
    return headers;
  } else {
    return {
      [authorizationHeaderName]: `Bearer ${config.authorizationKey}`,
    };
  }
};

const ensureNewContentHeader = (
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  const headerValue = `true`;
  if (headers) {
    headers[waitForLoadingNewContentHeaderName] = headerValue;
    return headers;
  } else {
    return {
      [waitForLoadingNewContentHeaderName]: headerValue,
    };
  }
};

const ensureTrackingHeader = (
  headers?: KontentHttpHeaders | undefined,
): KontentHttpHeaders => {
  const headerValue = `${packageName};${packageVersion}`;
  if (headers) {
    headers[trackingHeaderName] = headerValue;
    return headers;
  } else {
    return {
      [trackingHeaderName]: headerValue,
    };
  }
}

const loadAllKontentItems = async (
  config: CustomPluginOptions,
  language: string,
): Promise<KontentItemInput[]> => {
  let continuationToken = '';
  const items = [];
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  do {
    headers[continuationHeaderName] = continuationToken;

    const response = await axios.get(
      `${getProtocolAndDomain(config)}/${
        config.projectId
      }/items-feed?language=${language}${config?.experimental?.additionalItemFilterParams ? '&' + config.experimental.additionalItemFilterParams : ''}`,
      {
        headers,
        raxConfig: {
          onRetryAttempt: logRetryAttempt,
        },
      },
    );

    const union = _.unionBy<KontentItemInput>(
      response.data.items,
      Object.values(response.data.modular_content),
      'system.codename',
    );
    items.push(...union);
    continuationToken = response.headers[continuationHeaderName];
  } while (continuationToken);

  return items;
};

const loadKontentItem = async (
  itemId: string,
  language: string,
  config: CustomPluginOptions,
  waitForLoadingNewContent = false,
): Promise<{
  item: KontentItemInput | undefined;
  modularKontent: { [key: string]: KontentItemInput };
}> => {

  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  if (waitForLoadingNewContent) {
    headers = ensureNewContentHeader(headers)
  }

  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId
    }/items?system.id=${itemId}&language=${language}${config?.experimental?.additionalItemFilterParams ? '&' + config.experimental.additionalItemFilterParams : ''}`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );

  return {
    item: response.data.items.length > 0 ? response.data.items[0] : undefined,
    modularKontent: response.data.modular_content
  };
};

const loadAllKontentTypes = async (
  config: CustomPluginOptions,
): Promise<KontentType[]> => {
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId}/types`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );
  return response.data.types;
};

const loadAllKontentTypesCached = async (
  config: CustomPluginOptions,
  cache: GatsbyCache
): Promise<KontentType[]> => {
  let types = await cache.get(getKontentTypesCacheKey());
  if (!types) {
    types = await loadAllKontentTypes(config);
    cache.set(getKontentTypesCacheKey(), types)
  }

  return types;
}

const loadAllKontentTaxonomies = async (
  config: CustomPluginOptions,
): Promise<KontentTaxonomy[]> => {
  let headers = ensureAuthorizationHeader(config);
  headers = ensureTrackingHeader(headers);
  const response = await axios.get(
    `${getProtocolAndDomain(config)}/${config.projectId}/taxonomies`,
    {
      headers,
      raxConfig: {
        onRetryAttempt: logRetryAttempt,
      },
    },
  );
  return response.data.taxonomies;
};

export { loadKontentItem, loadAllKontentItems, loadAllKontentTypes, loadAllKontentTypesCached, loadAllKontentTaxonomies };
