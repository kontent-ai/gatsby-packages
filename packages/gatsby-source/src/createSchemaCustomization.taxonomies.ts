import { CreateSchemaCustomizationArgs } from 'gatsby';
import { getKontentTaxonomiesSchemaNamingConfiguration } from './naming';

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
): Promise<void> => {
  const schema = getKontentTaxonomiesSchemaNamingConfiguration();
  api.actions.createTypes(schema);
};

export { createSchemaCustomization as kontentTaxonomiesCreateSchemaCustomization };
