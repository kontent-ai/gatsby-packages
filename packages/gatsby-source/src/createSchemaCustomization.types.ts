import { CreateSchemaCustomizationArgs } from 'gatsby';
import { getKontentTypesSchemaNamingConfiguration } from './naming';

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
): Promise<void> => {
  const schema = getKontentTypesSchemaNamingConfiguration();
  api.actions.createTypes(schema);
};

export { createSchemaCustomization as kontentTypesCreateSchemaCustomization };
