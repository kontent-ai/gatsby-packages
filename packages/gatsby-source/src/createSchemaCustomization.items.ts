import { CreateSchemaCustomizationArgs } from 'gatsby';
import {
  CustomPluginOptions,
  KontentItemInput,
  KontentType,
  KontentTypeElementsObject,
  KontentTypeItemObjectDefinition,
} from './types';
import { loadAllKontentTypesCached } from './client';

import {
  getKontentItemElementsSchemaTypeName,
  getKontentItemNodeTypeName,
  getKontentItemsSchemaNamingConfiguration,
  getKontentItemSystemElementTypeName,
  getKontentItemInterfaceName,
  getKontentItemElementTypeNameByType,
  getKontentItemLanguageLinkExtensionName,
  PREFERRED_LANGUAGE_IDENTIFIER,
} from './naming';

const getLanguageLinkExtension = (): object => ({
  name: getKontentItemLanguageLinkExtensionName(),
  extend: (): object => ({
    async resolve(
      source: { value?: string[]; modular_content?: string[]; type: string },
      _args: unknown,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      context: any,
    ): Promise<KontentItemInput[]> {
      const nodesCodeNames =
        source.type === 'modular_content'
          ? (source.value as string[])
          : (source.modular_content as string[]);

      if (nodesCodeNames.length === 0) {
        return [];
      }

      const kontentItemNode = context.nodeModel.findRootNodeAncestor(source);
      const nodesLanguage = kontentItemNode[PREFERRED_LANGUAGE_IDENTIFIER];

      const promises = nodesCodeNames.map(codename =>
        context.nodeModel.findOne({
          query: {
            filter: {
              system: {
                codename: {
                  eq: codename,
                },
              },
              [PREFERRED_LANGUAGE_IDENTIFIER]: {
                eq: nodesLanguage,
              },
            },
          },
          type: getKontentItemInterfaceName(),
        }),
      );

      const nodes = await Promise.all(promises);
      return nodes;
    },
  }),
});

const getElementFieldsDefinitionForType = (
  type: KontentType,
): { [key: string]: { type: string } } => {
  const elementFields: {
    [key: string]: {
      type: string;
    };
  } = {};

  for (const elementKey in type.elements) {
    if (Object.prototype.hasOwnProperty.call(type.elements, elementKey)) {
      const element = (type.elements as KontentTypeElementsObject)[elementKey];
      const elementType = getKontentItemElementTypeNameByType(element.type);
      if (elementType !== '') {
        elementFields[elementKey] = {
          type: elementType,
        };
      }
    }
  }

  return elementFields;
};

const createSchemaCustomization = async (
  api: CreateSchemaCustomizationArgs,
  pluginConfig: CustomPluginOptions,
): Promise<void> => {
  // TODO check https://github.com/gatsbyjs/gatsby/pull/14610/files/5c50c435ab49884b6d854cd07f20efd95d1e5f52#diff-29de3acf9ce1010435f2b2f0043dba8cR252
  // failing for update run
  const languageExtension = getLanguageLinkExtension();
  api.actions.createFieldExtension(languageExtension);

  const baseSchemaTypes = getKontentItemsSchemaNamingConfiguration();
  api.actions.createTypes(baseSchemaTypes);

  const types = await loadAllKontentTypesCached(pluginConfig, api.cache);
  for (const type of types) {
    const kontentItemElementsTypeName = getKontentItemElementsSchemaTypeName(
      type.system.codename,
    );

    const elementsFieldDefinition = getElementFieldsDefinitionForType(type)
    const typeContainsElements = Object.keys(elementsFieldDefinition).length > 0;
    if (typeContainsElements) {
      const elementsTypeDef = api.schema.buildObjectType({
        name: kontentItemElementsTypeName,
        fields: elementsFieldDefinition,
        extensions: {
          infer: false
        }
      });
      api.actions.createTypes(elementsTypeDef);
    }

    const typeName = getKontentItemNodeTypeName(type.system.codename);
    const systemElementsTypeName = getKontentItemSystemElementTypeName();
    const typeInterfaceName = getKontentItemInterfaceName();
    const typeItemObjectDefinition: KontentTypeItemObjectDefinition = {
      name: typeName,
      fields: {
        system: `${systemElementsTypeName}!`,
        elements: kontentItemElementsTypeName,
        [PREFERRED_LANGUAGE_IDENTIFIER]: 'String!',
      },
      interfaces: ['Node', typeInterfaceName],
      infer: false,
    };
    if (!typeContainsElements) {
      delete typeItemObjectDefinition.fields.elements;
    }
    const typeItemDef = api.schema.buildObjectType(typeItemObjectDefinition);
    api.actions.createTypes(typeItemDef);
  }
};

export { createSchemaCustomization as kontentItemsCreateSchemaCustomization };
