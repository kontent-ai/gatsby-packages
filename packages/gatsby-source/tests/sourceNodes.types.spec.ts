import { kontentTypesSourceNodes } from '../src/sourceNodes.types';
import { SourceNodesArgs, Actions } from 'gatsby';
import { CustomPluginOptions, KontentType } from '../src/types';
import { createMock } from 'ts-auto-mock';
import { createContentDigest } from 'gatsby-core-utils';

// TODO fix lint error
import * as complexTypesFakeResponse from './complexTypesFakeResponse.json';

import { mocked } from 'ts-jest/dist/util/testing';
import * as _ from 'lodash';
import axios from 'axios';
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('sourceNodes', () => {
  describe('complex types', () => {
    const api = createMock<SourceNodesArgs>({
      createNodeId: jest.fn(input => `dummyId-${input}`),
      actions: createMock<Actions>({
        createNode: jest.fn(),
      }),
      createContentDigest,
    });

    const pluginConfiguration = createMock<CustomPluginOptions>({
      projectId: 'dummyProject',
    });

    mockedAxios.get.mockImplementation(url => {
      if (url.includes('types')) {
        return Promise.resolve({
          data: complexTypesFakeResponse,
          headers: [],
        });
      } else {
        throw new Error('Language should be defined in the language parameter');
      }
    });

    it('import all types correctly', async () => {
      await kontentTypesSourceNodes(api, pluginConfiguration);
      const createNodesMock = mocked(api.actions.createNode, true);
      const createdNodes = _.flatMap(
        createNodesMock.mock.calls,
      ) as KontentType[];

      expect(createdNodes.length).toBe(6);
      expect(createdNodes).toMatchSnapshot();
    });
  });
});
