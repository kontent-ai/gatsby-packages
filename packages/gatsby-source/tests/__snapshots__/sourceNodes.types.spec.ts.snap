// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sourceNodes complex types import all types correctly 1`] = `
Array [
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "profile_photo",
        "name": "Profile Photo",
        "type": "asset",
      },
      Object {
        "codename": "friends",
        "name": "Friends",
        "type": "modular_content",
      },
      Object {
        "codename": "category",
        "name": "Category",
        "taxonomy_group": "category",
        "type": "taxonomy",
      },
      Object {
        "codename": "custom_element",
        "name": "custom element",
        "type": "custom",
      },
      Object {
        "codename": "bio",
        "name": "Bio",
        "type": "rich_text",
      },
      Object {
        "codename": "name_and_surname",
        "name": "Name and Surname",
        "type": "text",
      },
      Object {
        "codename": "gender",
        "name": "Gender",
        "options": Array [
          Object {
            "codename": "man",
            "name": "Man",
          },
          Object {
            "codename": "woman",
            "name": "Woman",
          },
          Object {
            "codename": "other",
            "name": "Other",
          },
        ],
        "type": "multiple_choice",
      },
      Object {
        "codename": "favorite_number",
        "name": "Favorite Number",
        "type": "number",
      },
      Object {
        "codename": "url_slug",
        "name": "URL slug",
        "type": "url_slug",
      },
      Object {
        "codename": "date_of_birth",
        "name": "Date of Birth",
        "type": "date_time",
      },
    ],
    "id": "dummyId-kontent_type_person",
    "internal": Object {
      "contentDigest": "358dfd89e48ae7c223d32a3943f343af",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "person",
      "id": "081a7b99-575d-473c-9cd0-67094803da73",
      "last_modified": "2020-03-17T10:59:15.7647305Z",
      "name": "Person",
    },
  },
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "description",
        "name": "Description",
        "type": "custom",
      },
      Object {
        "codename": "name",
        "name": "Name",
        "type": "text",
      },
      Object {
        "codename": "url",
        "name": "URL",
        "type": "text",
      },
    ],
    "id": "dummyId-kontent_type_repository",
    "internal": Object {
      "contentDigest": "681b4d4d1954461316054c827e853aaf",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "repository",
      "id": "c88be77c-83a6-4839-855f-66c28f624d1d",
      "last_modified": "2019-04-01T18:29:04.7184530Z",
      "name": "Repository",
    },
  },
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "summary",
        "name": "Summary",
        "type": "rich_text",
      },
      Object {
        "codename": "name",
        "name": "Name",
        "type": "text",
      },
      Object {
        "codename": "assets",
        "name": "Assets",
        "type": "asset",
      },
      Object {
        "codename": "excercise",
        "name": "Excercise",
        "type": "rich_text",
      },
      Object {
        "codename": "agenda",
        "name": "Agenda",
        "type": "modular_content",
      },
      Object {
        "codename": "checklist",
        "name": "Checklist",
        "options": Array [
          Object {
            "codename": "training_sessions_scheduled",
            "name": "Training sessions scheduled",
          },
          Object {
            "codename": "training_rooms_reserved",
            "name": "Training rooms reserved",
          },
          Object {
            "codename": "presentations_and_files_proofread",
            "name": "Presentations and files proofread",
          },
          Object {
            "codename": "introduction_and_supporting_files_uploaded",
            "name": "Introduction and supporting files uploaded",
          },
          Object {
            "codename": "start_code_files_and_final_code_files_uploaded",
            "name": "START code files and FINAL code files uploaded",
          },
          Object {
            "codename": "exercise_step_by_step_instructions_added",
            "name": "Exercise step-by-step instructions added",
          },
          Object {
            "codename": "training_session_tagged_in_confluence",
            "name": "Training session tagged in Confluence",
          },
        ],
        "type": "multiple_choice",
      },
      Object {
        "codename": "presenter",
        "name": "Presenter",
        "type": "modular_content",
      },
    ],
    "id": "dummyId-kontent_type_session",
    "internal": Object {
      "contentDigest": "e6132a453367640f1229b0629120d319",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "session",
      "id": "9dfcd879-0d06-45ca-9b2c-67fd9c6e3b34",
      "last_modified": "2019-04-01T18:40:04.7086396Z",
      "name": "Session",
    },
  },
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "text",
        "name": "Text",
        "type": "rich_text",
      },
    ],
    "id": "dummyId-kontent_type_step",
    "internal": Object {
      "contentDigest": "16619b1849fc9cbc2f7bd93b30e6219c",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "step",
      "id": "c4b8e601-77f5-4535-b522-8b38db55873e",
      "last_modified": "2019-04-01T18:40:11.1952697Z",
      "name": "Step",
    },
  },
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "author",
        "name": "Author",
        "type": "modular_content",
      },
      Object {
        "codename": "name",
        "name": "Name",
        "type": "text",
      },
      Object {
        "codename": "sessions",
        "name": "Sessions",
        "type": "modular_content",
      },
      Object {
        "codename": "description",
        "name": "Description",
        "type": "rich_text",
      },
    ],
    "id": "dummyId-kontent_type_training",
    "internal": Object {
      "contentDigest": "0c9dc4b6daa8afcc390a192124de9442",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "training",
      "id": "d4cf750f-61d3-4782-bf9c-b7184b7b2fc3",
      "last_modified": "2019-04-01T18:44:00.4246990Z",
      "name": "Training",
    },
  },
  Object {
    "children": Array [],
    "elements": Array [
      Object {
        "codename": "url",
        "name": "URL",
        "type": "text",
      },
      Object {
        "codename": "name",
        "name": "Name",
        "type": "text",
      },
      Object {
        "codename": "description",
        "name": "Description",
        "type": "rich_text",
      },
    ],
    "id": "dummyId-kontent_type_website",
    "internal": Object {
      "contentDigest": "190c93492e9939118744998368beaffe",
      "type": "kontent_type",
    },
    "system": Object {
      "codename": "website",
      "id": "aeabe925-9221-4fb2-bc3a-2a91abc904fd",
      "last_modified": "2019-04-01T18:33:45.0353591Z",
      "name": "Website",
    },
  },
]
`;
